woocommerce/woocommerce-ios
# REST API: Fetch Jetpack connection status before starting the Jetpack setup flow

Closes: #8914 


## Description

This PR adds the Jetpack connection check to the Jetpack setup flow for non-Jetpack sites. The logic for this connection check can be found in an internal document here: [pe5sF9-1c9-p2].

Since the logic for showing the benefit modal is currently kept in `DashboardViewController` which is not scalable, I moved this logic to a new coordinator `JetpackSetupCoordinator` which takes care of showing the modal and handling the actions for both JCP sites and non-Jetpack sites:
- For JCP sites, the logic is kept as-is: when tapping the Install Jetpack on the modal, we dismiss the modal and present the installation flow for JCP sites.
- For non-Jetpack sites: tapping the Install Jetpack button triggers fetching Jetpack connection details while showing a loading indicator on the button. When the result returns, it is sent back to the coordinator to continue the setup flow - which will be handled in subsequent PRs.

Previously, `JetpackConnectionRemote` was designed to handle requests with cookie-nonce authentication only. To make use of this remote with application password authentication,  `JetpackConnectionStore` has been updated to be initialized with `AlamofireNetwork` in `AuthenticatedState`. This makes it possible to use `JetpackSetupViewModel` later without having to switch between `JetpackConnectionStore` and `SitePluginStore`.

## Testing instructions

Prerequisite: Make sure that you have access to a site with WooCommerce and no Jetpack.
- Log out of the app or skip onboarding if needed.
- Tap "Enter your site address" on the prologue screen and proceed with the address of your test site.
- Log in with the credentials of a shop manager.
- When the login completes, tap the Jetback benefit banner at the bottom of the Dashboard screen.
- Tap Install Jetpack on the benefit modal.
- Notice in Xcode console: `⛔️ Jetpack status fetched error: responseValidationFailed(reason: Alamofire.AFError.ResponseValidationFailureReason.unacceptableStatusCode(code: 404))`
- Install and activate Jetpack to your test site on WP Admin.
- Come back to the app and tap Install Jetpack again. You should see in Xcode console: `⛔️ Jetpack status fetched error: responseValidationFailed(reason: Alamofire.AFError.ResponseValidationFailureReason.unacceptableStatusCode(code: 403))`.
- Log out and log in again to an admin account of your test site.
- Try to tap the Install Jetpack on the Jetpack benefit modal again. Notice in Xcode console: `✅ connected email: nil, isConnected: false`.
- Use your admin user to handle Jetpack connection in WP Admin.
- Try again to tap the Install Jetpack button on the app. Notice in Xcode console: `✅ connected email: Optional("<your email>"), isConnected: true`

Feel free to also test the JCP case again:
- Deactivate Jetpack plugin on your site in WP Admin.
- Install and activate WCPay on your site. Connect your WPCom account to WCPay.
- Log out of the app and log in again to your site with your WPCom account.
- Tap the Jetpack benefit banner and then the Install Jetpack button on the benefit modal. The app should navigate to the installation flow for JCP sites.

## Screenshots

No new UI changes.

---
